src/drivers/dw/uart.c

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (c) 2017-2018, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include <sof/cpu.h>
+#include <sof/dw-uart.h>
+#include <sof/io.h>
+#include <sof/sof.h>
+#include <sof/string.h>
+#include <sof/wait.h>
+
+#include <platform/platform.h>
+
+#include "uart-priv.h"
+
+struct dw_uart_device_full {
+	struct dw_uart_device common;
+	uint32_t irq;
+	completion_t complete;
+};
+
+static struct dw_uart_device_full uart_dev = {
+	.common = {
+		.port = PLATFORM_LL_UART_REG_BASEADDR,
+		.retry = PLATFORM_LL_UART_RETRY,
+	},
+	/* Keep the CPU field 0 to be able to set it to the CPU, we're running on */
+	.irq = IRQ_EXT_HOST_UART(0),
+};
+
+/* Actually the FIFO size can be read out */
+#define DW_UART_FIFO_SIZE	64
+/* Using a ring buffer only makes sense when using a TX underrun IRQ */
+#define DW_UART_RING_SIZE 4096
+
+static uint8_t *dw_uart_ring;
+static bool dw_uart_ring_empty = true;
+/* Write to head, read from tail */
+static unsigned int dw_uart_ring_head;
+static unsigned int dw_uart_ring_tail;
+/* Protect the ring */
+static spinlock_t dw_uart_ring_lock;
```

Comment:
These should all be in a struct dw_uart.

Reply:
@lgirdwood speaking of which: do we need a device object to pass it around? We'd need a global `struct uart *trace_uart` pointer then.